File: app/src/main/java/com/yawar/memo/utils/BaseApp.java
package com.yawar.memo.utils;


import android.annotation.SuppressLint;
import android.app.Application;
import android.content.Intent;
import android.graphics.Color;
import android.os.Handler;
import android.text.TextUtils;
import android.util.Log;

import androidx.appcompat.app.AppCompatDelegate;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.lifecycle.ProcessLifecycleOwner;
import androidx.preference.PreferenceManager;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.Volley;
import com.google.firebase.auth.PhoneAuthProvider;
import com.yawar.memo.R;
import com.yawar.memo.observe.ContactNumberObserve;
import com.yawar.memo.observe.FireBaseTokenObserve;
import com.yawar.memo.observe.StoriesObserve;
import com.yawar.memo.repositry.AuthRepo;
import com.yawar.memo.repositry.BlockUserRepo;
import com.yawar.memo.repositry.ChatMessageRepo;
import com.yawar.memo.repositry.ChatRoomRepo;
import com.yawar.memo.repositry.RequestCallRepo;
import com.yawar.memo.repositry.UserInformationRepo;
import com.yawar.memo.service.SocketIOService;
import com.yawar.memo.sessionManager.ClassSharedPreferences;

public class BaseApp extends Application implements LifecycleObserver {

//    ChatRoomObserve observeClass;
    FireBaseTokenObserve fireBaseTokenObserve;
    StoriesObserve storiesObserve;
    ContactNumberObserve contactNumberObserve;
    public static final String TAG = "VolleyPatterns";
    private RequestQueue mRequestQueue;
    private static BaseApp sInstance;
    ChatRoomRepo chatRoomRepo;
    RequestCallRepo requestCallRepo;
    BlockUserRepo blockUserRepo;
    AuthRepo authRepo;
    ChatMessageRepo chatMessageRepo;
    UserInformationRepo userInformationRepo;
     String[] darkModeValues ;
     ClassSharedPreferences classSharedPreferences;
     Handler handler = new Handler();
    private Runnable myRunnable ;



    String peerId = null;




    @Override
    public void onCreate() {
        super.onCreate();
        setMode();




        sInstance = this;

        ProcessLifecycleOwner.get().getLifecycle().addObserver(this);
        System.out.println("getInstance()");
        if(getClassSharedPreferences().getUser()!=null) {
//            System.out.println("getChatRoomRepo().callAPI(classSharedPreferences.getUser().getUserId());");
            Log.d(TAG, "getChatRoomRepo().callAPI(classSharedPreferences.getUser()");
            getChatRoomRepo().callAPI(classSharedPreferences.getUser().getUserId());

        }




    }


     public  String isActivityVisible(){
         return ProcessLifecycleOwner.get().getLifecycle().getCurrentState().name();
     }
    @Override
    public void onTerminate() {
        super.onTerminate();
        System.out.println("on Tirminal");

    }

    @Override
    public void onLowMemory() {
        System.out.println("on low memoryyyyyyyyyyyyyy");
        super.onLowMemory();
    }

    @Override
    public void onTrimMemory(int level) {
        System.out.println("onTrimMemory BaseApp");

        super.onTrimMemory(level);
    }

    public static synchronized BaseApp getInstance() {
        if(sInstance==null){
            Log.d(TAG, "get Instance");
            sInstance = new BaseApp();
        }

        return sInstance;
        }

//        observeClass = new ChatRoomObserve();
//        fireBaseTokenObserve = new FireBaseTokenObserve();
//        storiesObserve = new StoriesObserve();
//        contactNumberObserve = new ContactNumberObserve();



//    public ChatRoomObserve getObserver() {
//       if(observeClass== null){
//           observeClass = new ChatRoomObserve();
//       }
//
//        return observeClass;
//    }
    public FireBaseTokenObserve getForceResendingToken() {
        if(fireBaseTokenObserve== null){
            fireBaseTokenObserve = new FireBaseTokenObserve();
        }
        return fireBaseTokenObserve;
    }
    public StoriesObserve getStoriesObserve() {
        if(storiesObserve== null){
            storiesObserve = new StoriesObserve();
        }
        return storiesObserve;
    }
    public ContactNumberObserve getContactNumberObserve() {
        if(contactNumberObserve== null){
            contactNumberObserve = new ContactNumberObserve();
        }
        return contactNumberObserve;
    }
    public RequestQueue getRequestQueue() {
        if (mRequestQueue == null) {
            mRequestQueue = Volley.newRequestQueue(getApplicationContext());
        }

        return mRequestQueue;
    }
    public ChatRoomRepo getChatRoomRepo() {
        if(chatRoomRepo== null){
            chatRoomRepo = new ChatRoomRepo(this);
//            if(getClassSharedPreferences().getUser()!=null)
//            chatRoomRepo.callAPI(classSharedPreferences.getUser().getUserId());

        }
        return chatRoomRepo;
    }
    public RequestCallRepo getRequestCallRepo() {
        if(requestCallRepo== null){
            requestCallRepo = new RequestCallRepo(this);
        }
        return requestCallRepo;
    }
    public UserInformationRepo getUserInformationRepo() {
        if(userInformationRepo== null){
            userInformationRepo = new UserInformationRepo(this);
        }
        return userInformationRepo;
    }
    public BlockUserRepo getBlockUserRepo() {
        if(blockUserRepo== null){
            blockUserRepo = new BlockUserRepo(this);
        }
        return blockUserRepo;
    }
    public AuthRepo getAuthRepo() {
        if(authRepo== null){
            authRepo = new AuthRepo(this);
        }
        return authRepo;
    }
    public ChatMessageRepo getChatMessageRepo() {
        if(chatMessageRepo== null){
            chatMessageRepo = new ChatMessageRepo(this);
        }
        return chatMessageRepo;
    }

    public ClassSharedPreferences getClassSharedPreferences() {
        if(classSharedPreferences== null){
            classSharedPreferences = new ClassSharedPreferences(this);
        }
        return classSharedPreferences;
    }
    public void setPeerId(String peer_id) {
        if(peerId== null){
            this.peerId = peer_id;
        }
    }
    public String getPeerId() {
    return  peerId;
    }

    public <T> void addToRequestQueue(Request<T> req, String tag) {
        req.setTag(TextUtils.isEmpty(tag) ? TAG : tag);
        getRequestQueue().add(req);
    }

    public <T> void addToRequestQueue(Request<T> req) {
        req.setTag(TAG);
        getRequestQueue().add(req);
    }
    public  void setMode(){
        darkModeValues = getResources().getStringArray(R.array.dark_mode_values);
        String pref = PreferenceManager.getDefaultSharedPreferences(this)
                .getString(getString(R.string.dark_mode), getString(R.string.dark_mode_def_value));
        // Comparing to see which preference is selected and applying those theme settings
        if (pref.equals(darkModeValues[0])){
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);}
        else if (pref.equals(darkModeValues[1])){

            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);}

    }

    public void cancelPendingRequests(Object tag) {
        if (mRequestQueue != null) {
            mRequestQueue.cancelAll(tag);
        }
    }
    @SuppressLint("CheckResult")
    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    public void onMoveToForeground() {


        if(classSharedPreferences.getUser()!=null){
//            if(chatR) {
//                getChatRoomRepo().callAPI(classSharedPreferences.getUser().getUserId());



        Intent service = new Intent(this, SocketIOService.class);
        service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_JOIN);
        this.startService(service);}


//        notifyAll();


    }

    @SuppressLint("CheckResult")
    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    public void onMoveToBackground() {
        if(classSharedPreferences.getUser()!=null) {


            Intent service = new Intent(this, SocketIOService.class);
            service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_DISCONNECT);
            startService(service);
//            myRunnable = new Runnable() {
//                @Override
//                public void run() {
//                    System.out.println("on Move to bacground");
//                    service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_DISCONNECT);
//                    startService(service);
//                }
//            };
//            handler.postDelayed(myRunnable, 30000);
        }




    }
    @SuppressLint("CheckResult")
    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    public void onDestroy() {
        System.out.println("on ON_DESTROY my App");
//            Intent service = new Intent(this, SocketIOService.class);
//            service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_DISCONNECT);
//            this.startService(service);



    }


}

File: app/src/main/java/com/yawar/memo/BaseApp.kt
package com.yawar.memo

//import com.yawar.memo.repositry.ChatRoomRepoo;
//import com.yawar.memo.repositry.RequestCallRepo;
import android.annotation.SuppressLint
import android.app.Application
import android.content.Intent
import android.os.Build
import android.os.Handler
import android.text.TextUtils
import android.util.Log
import androidx.appcompat.app.AppCompatDelegate
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.rxjava2.RxPreferenceDataStoreBuilder
import androidx.datastore.rxjava2.RxDataStore
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.OnLifecycleEvent
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.preference.PreferenceManager
import androidx.work.*
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.toolbox.Volley
import com.yawar.memo.observe.FireBaseTokenObserve
import com.yawar.memo.repositry.AuthRepo
import com.yawar.memo.repositry.BlockUserRepo
import com.yawar.memo.repositry.ChatMessageRepoo
import com.yawar.memo.repositry.ChatRoomRepoo
import com.yawar.memo.service.SocketIOService
import com.yawar.memo.sessionManager.ClassSharedPreferences
import com.yawar.memo.work.RefreshChatRoomWorker
import dagger.hilt.android.HiltAndroidApp
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.concurrent.TimeUnit
import javax.inject.Inject
@HiltAndroidApp
class BaseApp : Application(), LifecycleObserver {
    lateinit var  instance: BaseApp
    var fireBaseTokenObserve: FireBaseTokenObserve? = null
    private var mRequestQueue: RequestQueue? = null
    val applicationScope = CoroutineScope(Dispatchers.Default)
    @Inject
    lateinit var blockUserRepo: BlockUserRepo
    @Inject
    lateinit var authRepo: AuthRepo
    @Inject
    lateinit var chatMessageRepoo: ChatMessageRepoo
    lateinit var darkModeValues: Array<String>
    var classSharedPreferences: ClassSharedPreferences? = null
    var handler = Handler()
    private val myRunnable: Runnable? = null
    private var dataStore: RxDataStore<Preferences>? = null
    @Inject
    lateinit var chatRoomRepoo: ChatRoomRepoo
    private var peerId: String = ""
    override fun onCreate() {
        super.onCreate()
        setMode()
        sInstance = this
        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
        if (getClassSharedPreferences().user != null) {
//            chatRoomRepoo!!.loadChatRoom(classSharedPreferences!!.user.userId!!)
            delayedInit()
            Log.d(TAG, "onCreate: ")
        }
    }

    private fun delayedInit() = applicationScope.launch {
//        setupRecurringWork()
        chatRoomRepoo!!.loadChatRoom(classSharedPreferences!!.user.userId!!)
    }
//    private fun setupRecurringWork() {
//
//        val constraints = Constraints.Builder()
//            .setRequiredNetworkType(NetworkType.UNMETERED)
//            .setRequiresBatteryNotLow(true)
//            .setRequiresCharging(true)
//            .apply {
//                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
//                    setRequiresDeviceIdle(true)
//                }
//            }.build()
//        val repeatingRequest = PeriodicWorkRequestBuilder<RefreshChatRoomWorker>(1, TimeUnit.MICROSECONDS)
//            .setConstraints(constraints)
//            .build()
//
//        WorkManager.getInstance().enqueueUniquePeriodicWork(
//            RefreshChatRoomWorker.WORK_NAME,
//            ExistingPeriodicWorkPolicy.KEEP,
//            repeatingRequest)
//    }

    fun isActivityVisible(): String {
        return ProcessLifecycleOwner.get().lifecycle.currentState.name
    }

    override fun onTerminate() {
        super.onTerminate()
    }

    override fun onLowMemory() {
        super.onLowMemory()
    }

    override fun onTrimMemory(level: Int) {
        super.onTrimMemory(level)
    }

    fun getForceResendingToken(): FireBaseTokenObserve {
        if (fireBaseTokenObserve == null) {
            fireBaseTokenObserve = FireBaseTokenObserve()
        }
        return fireBaseTokenObserve!!
    }

    fun getRequestQueue(): RequestQueue {
        if (mRequestQueue == null) {
            mRequestQueue = Volley.newRequestQueue(applicationContext)
        }
        return mRequestQueue!!
    }

    @JvmName("getClassSharedPreferences1")
    fun getClassSharedPreferences(): ClassSharedPreferences {
        if (classSharedPreferences == null) {
            classSharedPreferences = ClassSharedPreferences(this)
        }
        return classSharedPreferences!!
    }

    fun getDataStore(): RxDataStore<Preferences> {
        if (dataStore == null) {
            dataStore = RxPreferenceDataStoreBuilder(this,  /*name=*/"settings").build()
        }
        return dataStore!!
    }

    fun setPeerId(peer_id: String?) {
        if (peerId == null) {
            peerId = peer_id!!
        }
    }

    fun <T> addToRequestQueue(req: Request<T>, tag: String?) {
        req.tag = if (TextUtils.isEmpty(tag)) TAG else tag
        getRequestQueue().add(req)
    }

    fun <T> addToRequestQueue(req: Request<T>) {
        req.tag = TAG
        getRequestQueue().add(req)
    }

    fun setMode() {
        darkModeValues = resources.getStringArray(R.array.dark_mode_values)
        val pref = PreferenceManager.getDefaultSharedPreferences(this)
            .getString(getString(R.string.dark_mode), getString(R.string.dark_mode_def_value))
        // Comparing to see which preference is selected and applying those theme settings
        if (pref == darkModeValues[0]) {
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
        } else if (pref == darkModeValues[1]) {
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        }
    }

    fun cancelPendingRequests(tag: Any?) {
        if (mRequestQueue != null) {
            mRequestQueue!!.cancelAll(tag)
        }
    }

    @SuppressLint("CheckResult")
    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    fun onMoveToForeground() {
        if (classSharedPreferences!!.user != null) {
            chatMessageRepoo!!.getUnRecivedMessages()
            val service = Intent(this, SocketIOService::class.java)
            service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_JOIN)
            startService(service)
        }
    }

    @SuppressLint("CheckResult")
    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    fun onMoveToBackground() {
        if (classSharedPreferences!!.user != null) {
            val service = Intent(this, SocketIOService::class.java)
            service.putExtra(
                SocketIOService.EXTRA_EVENT_TYPE,
                SocketIOService.EVENT_TYPE_DISCONNECT
            )
            startService(service)
        }
    }

    @SuppressLint("CheckResult")
    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroy() {
        Log.d(TAG, "onDestroy: ")
    }

    companion object {
        const val TAG = "BaseApp"
        private var sInstance: BaseApp? = null

        @get:Synchronized
        val instance: BaseApp?
            get() {
                if (sInstance == null) {
                    sInstance = BaseApp()
                }
                return sInstance
            }
    }
}

